File: Dynamic programming/codeforces_2050C.cpp

// Problem link ===>> https://codeforces.com/problemset/problem/2050/C
// submission link ===>> https://codeforces.com/contest/2050/submission/299478268

#include <bits/stdc++.h>
#define ll long long
using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    ll t;
    cin >> t;

    while (t--)
    {
        string n;
        cin >> n;

        int sum = 0, twos = 0, threes = 0;

        for (auto &ch : n)
        {
            sum += ch - '0';
            twos += ch == '2';
            threes += ch == '3';
        }

        bool found = false;

        for (int x = 0; x <= min(twos, 9); x++)
        {
            for (int y = 0; y <= min(threes, 3); y++)
            {
                if ((sum + 2 * x + 6 * y) % 9 == 0)
                {
                    cout << "YES\n";
                    found = true;
                    break;
                }
            }
            if (found)
                break;
        }

        if (!found)
        {
            cout << "NO\n";
        }
    }

    return 0;
}

File: Dynamic programming/codeforces_2050C.py

# Problem link ===>> https://codeforces.com/problemset/problem/2050/C


def main():
    t = int(input())
    for _ in range(t):
        s = [int(x) for x in list(input())]

        sm = sum(s)
        tows = s.count(2)
        threes = s.count(3)

        for i in range(min(10, tows+1)):
            for j in range(min(10, threes+1)):
                if (sm + i * 2 + j * 6) % 9 == 0:
                    print("YES")
                    return
        print("NO")


if __name__ == "__main__":
    main()
